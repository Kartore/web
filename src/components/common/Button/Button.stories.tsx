import { Button } from '.';
import { Meta, StoryObj } from '@storybook/react';
import { fn, userEvent, within, expect } from '@storybook/test';

const meta = {
  component: Button,
} satisfies Meta<typeof Button>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    children: 'Button',
  },
};

export const Loading: Story = {
  ...Default,
  args: {
    ...Default.args,
    isLoading: true,
  },
};

export const Disabled: Story = {
  ...Default,
  args: {
    ...Default.args,
    isDisabled: true,
  },
};

export const Click: Story = {
  ...Default,
  args: {
    ...Default.args,
    onClick: fn(),
  },
  play: async ({ args, canvasElement }) => {
    const canvas = within(canvasElement);
    const button = canvas.getByRole('button', {
      name: 'Button',
    });
    await userEvent.click(button);
    await expect(args.onClick).toHaveBeenCalledTimes(1);
  },
};
