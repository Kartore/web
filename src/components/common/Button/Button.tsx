import type { ComponentPropsWithoutRef, FC, MutableRefObject } from 'react';
import { useRef } from 'react';

import type { AriaButtonOptions } from 'react-aria';
import { useButton } from 'react-aria';

export type ButtonProps = AriaButtonOptions<'button'> &
  ComponentPropsWithoutRef<'button'> & {
    buttonRef?: MutableRefObject<HTMLButtonElement | null>;
  };

export const Button: FC<ButtonProps> = (props) => {
  const ref = useRef<HTMLButtonElement | null>(null);
  const { buttonRef = ref, ...otherProps } = props;
  const { buttonProps } = useButton(otherProps, buttonRef);
  return (
    <button {...buttonProps} ref={buttonRef} style={props.style}>
      {props.children}
    </button>
  );
};
