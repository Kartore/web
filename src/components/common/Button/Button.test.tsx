import { composeStories } from '@storybook/react';

import * as Stories from './Button.stories';
import { render } from '@testing-library/react';

const { Default, Loading, Disabled, Click } = composeStories(Stories);

describe('Component: Button', () => {
  describe('Snapshot', () => {
    it('Default', () => {
      const { asFragment } = render(<Default />);
      expect(asFragment()).toMatchSnapshot();
    });
    it('Loading', () => {
      const { asFragment } = render(<Loading />);
      expect(asFragment()).toMatchSnapshot();
    });
    it('Disabled', () => {
      const { asFragment } = render(<Disabled />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe('Interaction', () => {
    it('Click', async () => {
      const { container } = render(<Click />);
      await Click.play!({ canvasElement: container, args: Click.args });
    });
  });
});
