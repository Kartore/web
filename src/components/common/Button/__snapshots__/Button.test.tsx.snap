// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component: Button > Snapshot > Default 1`] = `
<DocumentFragment>
  <button
    class=""
    type="button"
  >
    Button
  </button>
</DocumentFragment>
`;
